cmd/sidecars/sidecar_shim.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	vmSchema "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/hooks"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	hooksV1alpha2 "kubevirt.io/kubevirt/pkg/hooks/v1alpha2"
+)
+
+const (
+	onDefineDomainLoggingMessage = "OnDefineDomain method has been called"
+
+	onDefineDomainBin = "onDefineDomain"
+)
+
+type infoServer struct {
+	Version string
+}
+
+func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*hooksInfo.InfoResult, error) {
+	log.Log.Info("Info method has been called")
+	return &hooksInfo.InfoResult{
+		Name: "shim",
+		Versions: []string{
+			s.Version,
+		},
+		HookPoints: []*hooksInfo.HookPoint{
+			{
+				Name:     hooksInfo.OnDefineDomainHookPointName,
+				Priority: 0,
+			},
+		},
+	}, nil
+}
+
+type v1Alpha1Server struct{}
+type v1Alpha2Server struct{}
+
+func (s v1Alpha2Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha2.OnDefineDomainParams) (*hooksV1alpha2.OnDefineDomainResult, error) {
+	log.Log.Info(onDefineDomainLoggingMessage)
+	if newDomainXML, err := runOnDefineDomain(params.GetVmi(), params.GetDomainXML()); err != nil {
+		return nil, err
+	} else {
+		return &hooksV1alpha2.OnDefineDomainResult{
+			DomainXML: newDomainXML,
+		}, nil
+	}
+}
+func (s v1Alpha2Server) PreCloudInitIso(_ context.Context, params *hooksV1alpha2.PreCloudInitIsoParams) (*hooksV1alpha2.PreCloudInitIsoResult, error) {
+	log.Log.Info("PreCloudInitIso method has been called")
+	return &hooksV1alpha2.PreCloudInitIsoResult{
+		CloudInitData: params.GetCloudInitData(),
+	}, nil
+}
+
+func (s v1Alpha1Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
```

Comment:
I'm not sure as I'm not proficient in Generics but I can't see how. For each version of v1alpha we might define the same method name `OnDefineDomain` but their `params` and `result` types might differ. I'm not sure if type inference would work in this case (e.g: using P and R for each type T, where P is for param type and R for the result type).

Again, not proficient in it but not sure it would be worth it.

We would still need to implement the methods in `v1alpha2` server that are not present in `v1alpha1` and the same for [v1alpha3](https://github.com/kubevirt/kubevirt/pull/10046)

All in all, if you know how to do it, please do tell!

Thanks for the review

Reply:
I see, it will make things too complicated for now. It was just an idea.